### DISCUSSION TITLE
[@types/sinon] TS2589 type instantiation is excessively deep when stubbing some methods

### DISCUSSION THREAD

**Discussion by abianco-doxee** [DefinitelyTyped/DefinitelyTyped#67156] [Issues with a @types package]
This error is reported when stubbing some methods of [mikro-orm](https://github.com/mikro-orm/mikro-orm) and trying to control the returned value through the `withArgs` method of the stub.

I created a "minimal" reproduction of the problem here https://github.com/alebianco-doxee/sinon-mikroorm-issue-stub-types while doing the initial investigation

[a comment by an automated account is not shown]

**Comment by 43081j** [comment]
this is because of recursive types in mikro-orm, im not entirely sure of a solution off the top of my head

here's a basic repro:

https://www.typescriptlang.org/play?experimentalDecorators=true#code/JYOwLgpgTgZghgYwgAgMqgPYgLJzAgC2mQG8AoZS5SAZzAAoA3OAGwC5k4QBPASg4BGGDCwhcA3GQC+ZMmG4AHFLnwEACnChhgrAIJQA5gFcAthHA0APABUAfMgC8pClQDaAaWShkAawjcMGGRrAF0Afg56aw8Q5AgAD0gQABMaZAwBACsIBDBkMOQVQg0tHRZ9YzMLGxj7Dmj3EN5JKUk5RRQAMWFHUhhhDjooUAMpZAAfZG6MVxDJeSUp4Qq0p1dpufbF6fQQLF6i9U1tPUNTczAraZXbcSA

it drills down to recursive types like `type Foo = string | Foo[];` being infinite.

i suppose we could change the type to stop when it sees an array somehow

**Comment by abianco-doxee** [comment]
my type-ing skills are not the best, but would this achieve the same effect without the recursion issue?

https://www.typescriptlang.org/play?experimentalDecorators=true#code/JYOwLgpgTgZghgYwgAgMqgPYgLJzAgC2mQG8AoZS5SAZzAAoA3OAGwC5k4QBPASg4BGGDCwhcA3GQC+ZMmG4AHFLnwEACnChhgrAIJQA5gFcAthHA0APABUAfMgC8ya8ggAPSCAAmNZPqhw3JagMMQAqrYUyAD8flABQSqEGlo6LPrGZhaWEfZRHC7unj7IGAIAVhAIYFGxJMgA2gDSyKDIANYQ3BgwzgC60RxJ6praeoam5mBW1s199jKUBbLySsgAYsKOpDDCHHRQoAZSyAA+G8INfZKrKJsYGb5ODffXcop3wuggWNvDKWN0hMstNLPdHrZxEA

**Reply by 43081j** [reply]
yup that makes sense to me 👍

**Reply by abianco-doxee** [reply]
would you need a PR to move the fix forward? and if not, can you tag me in the ticket/PR with the fix, just to keep track of when the fix lands

thanks

**Reply by 43081j** [reply]
if you have time to put a PR together, that'd be great. otherwise though i don't mind throwing one up some time this week

**Reply by abianco-doxee** [reply]
tried out that solution into the actual library types, unfortunately does not work as is. 
I'll try to tweak it a bit to make it work in the following days
